### ISSUE TITLE
[GDB] The "stepi" in arc-linux-gdb misbehaves.

### ISSUE THREAD

**Issue by shahab-vahedi** [foss-for-synopsys-dwc-arc-processors/toolchain#267]
version: `arc-2020.09`

The "stepi" command skip over a few instructions. It looks like it is miscalculation where the next possible instruction will be.

Compare the `arc-elf32-gdb`:
![gdbelf](https://user-images.githubusercontent.com/16524902/81323723-16309c00-9096-11ea-9111-b4e2104c4bb1.gif)

with `arc-linux-gdb` (with an extra warning on qemu side):
![gdblnx](https://user-images.githubusercontent.com/16524902/81323740-1c267d00-9096-11ea-97ca-56a58aa5d86a.gif)

[gdb-linux-stepi-tlbmissi.zip](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/4594577/gdb-linux-stepi-tlbmissi.zip)

**Comment by shahab-vahedi** [comment]
The divergence between `arc-elf32-gdb` and `arc-linux-gdb` is the existence of `arc_linux_software_single_step()` _gdbarch_ method in the latter:

```
arc-linux-gdb call stack:
     resume_1()
      maybe_software_singlestep()      --> returns 0 as "step"
       insert_single_step_breakpoints()
        gdbarch_software_single_step()
         arc_linux_software_single_step()

arc-elf32-gdb call stack:
     resume_1()
      maybe_software_singlestep()      --> returns 1 as "step"
```

`step=1` means going on with hw breakpoints, `step=0` means software breakpoints:
```
arc-linux-gdb:
infrun.c:resume_1()
{
  ...
  do_target_resume (resume_ptid, step=0, sig);
}

arc-elf32-gdb:
infrun.c:resume_1()
{
  ...
  do_target_resume (resume_ptid, step=1, sig);
}
```

I still need to unravel how the insertion of breakpoints and `step=0` contribute to this behavior.

**Comment by shahab-vahedi** [comment]
@anthony-kolesov kindly provided insights on this matter. Here comes the gist of it.

#### TLDR
Use `arc-elf32-gdb` for baremetal applications. This includes debugging the kernel.

#### Reason
The baremetal debugger (`arc-elf32-gdb`) has much more control over the program execution and its data. Among these lies the _hardware single step_. In this way, the debugger asks the server (`qemu`, `openocd`, etc.), by setting the appropriate bits, to stop at the very next **instruction**, whatever that may be. It can be the next linear instruction, the target of a jump or the beginning of an exception handler routine.

The linux user space debugger (`arc-linux-gdb`) assumes itself in a much constrained scenario where it can only ask the kernel's `ptrace` to set breakpoints for what it believes to be the next possible instruction. As an example, for a conditional branch the possible outcomes are the _target of the branch_ and the _fall-through_ (to not complicate this text more than necessary, _delay-slots_ are left out intentionally). Therefore, the linux user space debugger reacts in a much limited way. It does not even assume it should know about the kernel space exception handlers.

Of course there are ways to improve this, like moving the Single Step Instruction (SSI) logic from the debugger to the GDB server side. This way the debugger will only issue a SSI-RSP (Remote Serial Protocol) command and the server will take care of the rest. If it is baremetal application, the server end will use hardware stepping. If it is indeed served by a user-space server like _gdbserver_, then the same logic as discussed earlier could apply. However, the client would remain SSI logic agnostic.
Even better, the whole logic can be moved to kernel's ptrace for ARC. That way both user space debug servers known today, _gdbserver_ and _lldb-server_, could benefit from it.

Last but not least, I believe the _inconsistency_ warning that we see from QEMU's disassembly is rooted into the fact that the user space debugger is inserting software breakpoints (`SWI/SWI_S`) into the already disassembled code.
